### PR TITLE
Suggestion for "times after now"

### PR DESCRIPTION
**Pull request by akeranen**
One possible way to address the problem with positive relative times

### CODE REVIEW

File: draft-ietf-core-senml.md [hunk 1]
```diff
@@ -291,18 +291,19 @@ fields in each Record are the empty string so they are omitted.
...
-relative time in the past from the current time. A simple sensor with
-no absolute wall clock time might take a measurement every second,
-batch up 60 of them, and then send the batch to a server. It would
-include the relative time each measurement was made compared to the
-time the batch was sent in each SenML Record. The server might have
-accurate NTP time and use the time it received the data, and the
-relative offset, to replace the times in the SenML with absolute times
-before saving the SenML information in a document database.
+seconds. Values greater than 2^28 represent an absolute time relative
+to the Unix epoch. Values of 0 or less represent a relative time in
```
Comment (cabo) [inline draft-ietf-core-senml.md]: These are two cases, not three: one is relative to epoch, one is relative to "now".
Maybe we also should say that these are in seconds.

File: draft-ietf-core-senml.md [hunk 2]
```diff
@@ -291,18 +291,19 @@ fields in each Record are the empty string so they are omitted.
...
-"seconds since the epoch" {{TIME_T}}.  Values of 0 or less represent a
-relative time in the past from the current time. A simple sensor with
-no absolute wall clock time might take a measurement every second,
-batch up 60 of them, and then send the batch to a server. It would
-include the relative time each measurement was made compared to the
-time the batch was sent in each SenML Record. The server might have
-accurate NTP time and use the time it received the data, and the
-relative offset, to replace the times in the SenML with absolute times
-before saving the SenML information in a document database.
+seconds. Values greater than 2^28 represent an absolute time relative
```
Comment (cabo) [inline draft-ietf-core-senml.md]: Are we using ^ for exponentiation?  We probably need to point this out somewhere, as C uses ^ for XOR.
(I tend to use ** for exponentiation to keep ^ open for XOR usage.  See end of page 9 of RFC 7252 for a good way to do this.)

Comment (cabo) [inline draft-ietf-core-senml.md]: I would keep 2**28 in the POSIX case; having a closed interval is strange here.
... of 2**28 or greater ...

File: draft-ietf-core-senml.md [hunk 3]
```diff
@@ -291,18 +291,19 @@ fields in each Record are the empty string so they are omitted.
...
+to the Unix epoch. Values of 0 or less represent a relative time in
+the past from the current time. Values from 0 to 2^28 represent
```
Comment (cabo) [inline draft-ietf-core-senml.md]: "values less than 2**28", then.

File: draft-ietf-core-senml.md [hunk 4]
```diff
@@ -518,14 +519,27 @@ by the application context.
...
-of zero indicates that the sensor does not know the absolute time and
-the measurement was made roughly "now". A negative value is used to
-indicate seconds in the past from roughly "now". A positive value is
-used to indicate the number of seconds, excluding leap seconds, since
-the start of the year 1970 in UTC.
+If either the Base Time or Time value is missing, the missing field is
+considered to have a value of zero. The Base Time and Time values are
+added together to get the time of measurement. A time of zero
+indicates that the sensor does not know the absolute time and the
+measurement was made roughly "now". A negative value is used to
```
Comment (cabo) [inline draft-ietf-core-senml.md]: Again, this complicates things by talking about three cases.  There are only two: now-referenced and epoch-referenced.

File: draft-ietf-core-senml.md [hunk 5]
```diff
@@ -518,14 +519,27 @@ by the application context.
...
+indicate seconds in the past from roughly "now". 
+
+Positive values up to 2^28 indicate seconds in the future from "now".
```
Comment (cabo) [inline draft-ietf-core-senml.md]: (Example here is fine; but "less than 2**28" again.)

File: draft-ietf-core-senml.md [hunk 6]
```diff
@@ -518,14 +519,27 @@ by the application context.
...
+These can be used e.g., for actuation use when the desired change
+should happen in the future but the sender or the receiver does not
+have accurate time available.
+
+Values greater than 2^28 represent an absolute time relative to the
+Unix epoch (1970-01-01T00:00Z in UTC time) and the time is counted
+same way as the Portable Operating System Interface (POSIX) "seconds
+since the epoch" {{TIME_T}}.
```
Comment (cabo) [inline draft-ietf-core-senml.md]: Maybe mention Time.at(2**28).utc => 1978-07-04 21:24:16 UTC as the minimum absolute time value that can be expressed this way.

File: draft-ietf-core-senml.md [hunk 7]
```diff
@@ -518,14 +519,27 @@ by the application context.
...
+
+Because time values up to 2^28 are used for presenting time relative
+to "now" and Time and Base Time are added together, care must be taken
+to ensure that the sum does not inadvertently become greater and 2^28
+(i.e., absolute time) when relative time was intended to be used.
```
Comment (cabo) [inline draft-ietf-core-senml.md]: This sentence is weird.

**Review by cabo — CHANGES_REQUESTED** [review]

File: draft-ietf-core-senml.md (hunk 2 above)
Comment (fluffy) [reply draft-ietf-core-senml.md]: might be good to write this as    268,435,456 (2**28)

Comment (fluffy) [reply draft-ietf-core-senml.md]: I do think that 2^28 will be clear from context that we don't mean xor but I don't care about this one way or the other.

File: draft-ietf-core-senml.md (hunk 7 above)
Comment (akeranen) [reply draft-ietf-core-senml.md]: Any suggestions? The idea is to have the same disclaimer as before.

**Review by fluffy — APPROVED** [review]
Looks like two of you are still working details of phrasing but I like where it is going and all looks about right to me.

File: draft-ietf-core-senml.md (hunk 1 above)
Comment (akeranen) [reply draft-ietf-core-senml.md]: Changed here. The seconds are mentioned in the previous sentence.

File: draft-ietf-core-senml.md (hunk 2 above)
Comment (akeranen) [reply draft-ietf-core-senml.md]: Fixed.

Comment (akeranen) [reply draft-ietf-core-senml.md]: Makes sense. Fixed.

File: draft-ietf-core-senml.md (hunk 4 above)
Comment (akeranen) [reply draft-ietf-core-senml.md]: Re-wrote this too. First paragraph is about relative times and second absolute times. In the first paragraph still the sub-cases (before, now, after) are explained to make it clear.

File: draft-ietf-core-senml.md (hunk 5 above)
Comment (akeranen) [reply draft-ietf-core-senml.md]: Fixed

File: draft-ietf-core-senml.md (hunk 6 above)
Comment (akeranen) [reply draft-ietf-core-senml.md]: Fixed.

File: draft-ietf-core-senml.md [hunk 8]
```diff
@@ -521,24 +524,27 @@ by the application context.
...
-since the epoch" {{TIME_T}}.
-
-Because time values up to 2^28 are used for presenting time relative
+added together to get the time of measurement. 
+
+Values less than 2**28 represent time relative to the current time.
+That is, a time of zero indicates that the sensor does not know the
+absolute time and the measurement was made roughly "now". A negative
+value indicates seconds in the past from roughly "now". Positive
+values up to 2**28 indicate seconds in the future from "now". These
```
Comment (cabo) [inline draft-ietf-core-senml.md]: s/These/The latter/
(make reference clearer)

File: draft-ietf-core-senml.md (hunk 7 above)
Comment (cabo) [reply draft-ietf-core-senml.md]: maybe s/become greater and/become greater than/?
(Actually, I'd prefer "reach or exceed".)

File: draft-ietf-core-senml.md (hunk 8 above)
Comment (akeranen) [reply draft-ietf-core-senml.md]: This became "Positive values can be used..."

**Comment by akeranen** [comment]
@cabo Fixed

**Pull request merged**